### ISSUE TITLE
Docs: add missing XML documentation for sniffs

### ISSUE THREAD

**Issue by jrfnl** [PHPCSStandards/PHP_CodeSniffer#148]
Sniffs in PHP_CodeSniffer should preferably be accompanied by documentation. There are currently still a number of sniffs which don't have documentation.

Sniff documentation is provided via XML files in the standard's `Docs` directory and is available to end-users via the command-line and/or can be compiled into an HTML or Markdown file.

To see an example of some of some available documentation, run:
```bash
phpcs --standard=PSR12 --generator=Text
```

## Getting started

The [CONTRIBUTING](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/5a2a3769d2713de829a16a29c8bffad673eca6b5/.github/CONTRIBUTING.md#writing-sniff-documentation) doc contains information on writing sniff documentation and guidelines which should be followed.

### To Do

* [ ] `Generic.Arrays.ArrayIndent`
* [ ] `Generic.CodeAnalysis.EmptyPHPStatement`
* [ ] `Generic.Commenting.DocComment`
* [ ] `Generic.Formatting.SpaceBeforeCast`
* [ ] `Generic.PHP.RequireStrictTypes`
* [ ] `Generic.PHP.Syntax`
* [ ] `Generic.VersionControl.GitMergeConflict`
* [ ] `Generic.WhiteSpace.IncrementDecrementSpacing`
* [ ] `Generic.WhiteSpace.LanguageConstructSpacing`
* [ ] `PSR12.Classes.AnonClassDeclaration`
* [ ] `PSR12.Classes.ClosingBrace`
* [ ] `PSR12.Classes.OpeningBraceSpace`
* [ ] `PSR12.ControlStructures.BooleanOperatorPlacement`
* [ ] `PSR12.ControlStructures.ControlStructureSpacing`
* [ ] `PSR12.Files.DeclareStatement`
* [ ] `PSR12.Files.FileHeader`
* [ ] `PSR12.Files.ImportStatement`
* [ ] `PSR12.Files.OpenTag`
* [ ] `PSR12.Functions.ReturnTypeDeclaration`
* [ ] `PSR12.Properties.ConstantVisibility`
* [ ] `PSR12.Traits.UseDeclaration`
* [ ] `Squiz.Classes.ClassDeclaration`
* [ ] `Squiz.Classes.ClassFileName`
* [ ] `Squiz.Classes.ValidClassName`
* [ ] `Squiz.Commenting.BlockComment`
* [ ] `Squiz.Commenting.ClassComment`
* [ ] `Squiz.Commenting.ClosingDeclarationComment`
* [ ] `Squiz.Commenting.EmptyCatchComment`
* [ ] `Squiz.Commenting.FileComment`
* [ ] `Squiz.Commenting.FunctionComment`
* [ ] `Squiz.Commenting.InlineComment`
* [ ] `Squiz.Commenting.LongConditionClosingComment`
* [ ] `Squiz.Commenting.PostStatementComment`
* [ ] `Squiz.Commenting.VariableComment`
* [ ] `Squiz.ControlStructures.ControlSignature`
* [ ] `Squiz.ControlStructures.ElseIfDeclaration`
* [ ] `Squiz.ControlStructures.InlineIfDeclaration`
* [ ] `Squiz.ControlStructures.SwitchDeclaration`
* [ ] `Squiz.Files.FileExtension`
* [ ] `Squiz.Formatting.OperatorBracket`
* [ ] `Squiz.Functions.FunctionDeclarationArgumentSpacing`
* [ ] `Squiz.Functions.FunctionDeclaration`
* [ ] `Squiz.Functions.GlobalFunction`
* [ ] `Squiz.Functions.MultiLineFunctionDeclaration`
* [ ] `Squiz.NamingConventions.ValidFunctionName`
* [ ] `Squiz.NamingConventions.ValidVariableName`
* [ ] `Squiz.Objects.ObjectInstantiation`
* [ ] `Squiz.Operators.ComparisonOperatorUsage`
* [ ] `Squiz.Operators.IncrementDecrementUsage`
* [ ] `Squiz.Operators.ValidLogicalOperators`
* [ ] `Squiz.PHP.CommentedOutCode`
* [ ] `Squiz.PHP.DisallowBooleanStatement`
* [ ] `Squiz.PHP.DisallowComparisonAssignment`
* [ ] `Squiz.PHP.DisallowInlineIf`
* [ ] `Squiz.PHP.DisallowMultipleAssignments`
* [ ] `Squiz.PHP.DisallowSizeFunctionsInLoops`
* [ ] `Squiz.PHP.DiscouragedFunctions`
* [ ] `Squiz.PHP.EmbeddedPhp`
* [ ] `Squiz.PHP.Eval`
* [ ] `Squiz.PHP.GlobalKeyword`
* [ ] `Squiz.PHP.Heredoc`
* [ ] `Squiz.PHP.InnerFunctions`
* [ ] `Squiz.PHP.LowercasePHPFunctions`
* [ ] `Squiz.PHP.NonExecutableCode`
* [ ] `Squiz.Scope.MemberVarScope`
* [ ] `Squiz.Scope.MethodScope`
* [ ] `Squiz.Strings.ConcatenationSpacing`
* [ ] `Squiz.Strings.DoubleQuoteUsage`
* [ ] `Squiz.WhiteSpace.ControlStructureSpacing`
* [ ] `Squiz.WhiteSpace.FunctionClosingBraceSpace`
* [ ] `Squiz.WhiteSpace.FunctionOpeningBraceSpace`
* [ ] `Squiz.WhiteSpace.FunctionSpacing`
* [ ] `Squiz.WhiteSpace.LogicalOperatorSpacing`
* [ ] `Squiz.WhiteSpace.MemberVarSpacing`
* [ ] `Squiz.WhiteSpace.OperatorSpacing`
* [ ] `Squiz.WhiteSpace.PropertyLabelSpacing`
* [ ] `Squiz.WhiteSpace.ScopeClosingBrace`
* [ ] `Squiz.WhiteSpace.SuperfluousWhitespace`

Note: a number of sniffs will be removed in v 4.0. Those have been deliberately excluded from the above action list.

### Claimed/Work in Progress

_Nothing yet_

### Done

_Nothing yet_

## Want to contribute ?

Leave a comment below to claim a sniff you'll be working on.

PRs related to this task should preferably only contain the docs for one sniff each.

**Comment by rodrigoprimo** [comment]
> Leave a comment below to claim a sniff you'll be working on.

@jrfnl, I can help with this issue. I will start with the `Generic.Formatting.SpaceBeforeCast` sniff and, if all goes well, I can create PRs for more sniffs. I will make sure to create one PR per sniff.

Should we document the options that a given sniff supports? If so, how? For example, the `Generic.Arrays.ArrayIndent` supports `indent`, but I'm not sure how to document it by checking the instructions in CONTRIBUTING.md, and I was also not able to find any examples checking a few of the sniffs that are already documented.

**Comment by rodrigoprimo** [comment]
Sounds good to me. Thanks for the additional details.

**Comment by rodrigoprimo** [comment]
In the next few days, I can create PRs to add documentation to the `Generic` sniffs that are missing docs.

(please let me know if you prefer that I claim one sniff at a time instead of claiming a group of sniffs as I'm doing here)

**Comment by jrfnl** [comment]
> @jrfnl, in the next few days, I can create PRs to add documentation to the `Generic` sniffs that are missing docs.
> 
> (please let me know if you prefer that I claim one sniff at a time instead of claiming a group of sniffs as I'm doing here)

Great! Moved those to "claimed" now. Just make sure you submit each doc as a separate PR (but you knew that already).

**Comment by dingo-d** [comment]
I could take the PSR12 ones 👍🏼

**Comment by jrfnl** [comment]
> I could take the PSR12 ones 👍🏼

Moved to "Claimed" 👍🏻

**Comment by przemekhernik** [comment]
@jrfnl I want to learn how to create a proper sniff doc, so I can take care of `Squiz.WhiteSpace.FunctionClosingBraceSpace` as a first.

**Comment by jrfnl** [comment]
@przemekhernik 👍🏻 Moved to "Claimed"

**Comment by przemekhernik** [comment]
@jrfnl I have a small question about docs. Generally, I'm trying to use reverse engineering to check what specific sniff does to write docs based on the [sniff code](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/WhiteSpace/FunctionClosingBraceSpaceSniff.php). Sometimes I'm wondering if there some better way. What is your/recommended approach here?

For example, I'm not sure how to check [this part](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/WhiteSpace/FunctionClosingBraceSpaceSniff.php#L65-L80) of the sniff. Also, I've get through opened issues and find out that there's a [issue](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/6) for removing JS/CSS tokenizers. This one seems to be JS-one so should I bother myself about this in the docs?

**Comment by jrfnl** [comment]
> @jrfnl I have a small question about docs. Generally, I'm trying to use reverse engineering to check what specific sniff does to write docs based on the [sniff code](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/WhiteSpace/FunctionClosingBraceSpaceSniff.php). Sometimes I'm wondering if there some better way. What is your/recommended approach here?

@przemekhernik Well, best is obviously if the docs would (have been) written when the sniff was, but for this exercise, reverse engineering is unfortunately needed.

I can imagine different people having different approaches, but I would normally run the sniff over the test case file using the code report to see what is being flagged with which error code and combine that with reading the test case file taking note of what is **_not_** being flagged too + reading the sniff code.

For this particular sniff, the command I'd use for the test case file would be like so:
```bash
phpcs -ps ./src/Standards/Squiz/Tests/WhiteSpace/FunctionClosingBraceSpaceUnitTest.inc --standard=Squiz --report=code,source --sniffs=Squiz.WhiteSpace.FunctionClosingBraceSpace
```


> For example, I'm not sure how to check [this part](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/Squiz/Sniffs/WhiteSpace/FunctionClosingBraceSpaceSniff.php#L65-L80) of the sniff. Also, I've get through opened issues and find out that there's a [issue](https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/6) for removing JS/CSS tokenizers. This one seems to be JS-one so should I bother myself about this in the docs?

Feel free to ignore anything JS/CSS specific while writing the docs as, as you already saw, support for JS/CSS will be dropped soon enough, so just focusing on what the sniff does for PHP code should be fine.

**Comment by jrfnl** [comment]
@przemekhernik Just looked at the code and the test for that sniff and it looks like you unfortunately selected a sniff which has a [huge chunk of untested code](https://coveralls.io/builds/65500406/source?filename=src%2FStandards%2FSquiz%2FSniffs%2FWhiteSpace%2FFunctionClosingBraceSpaceSniff.php)..., so the tests vs code approach will not be sufficient in this case. 😞

**Comment by jrfnl** [comment]
@jonmcp Thank you. Added them to the "Claimed" list (and removed the PropertyLabel one - you are 100% correct, that one shouldn't have been on the list)

**Comment by jrfnl** [comment]
@dingo-d Thank you for the heads-up and thank you for the work already done on this. Docs for 8 out of 12 sniffs were merged so far, so that's great progress already!

I will move the remaining four back to "Unclaimed" and will close the PRs. If/when someone claims them and wants to continue on the work you've done for these already, the PR(s) can be reopened.

**Comment by stronk7** [comment]
> In docs for some external sniffs, I've sometimes mentioned that the behaviour is adjustable via properties, but I don't think that is done anywhere in this repo, so for now, I would leave it out. The wiki contains a page with details on all [customisable properties](https://github.com/PHPCSStandards/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties) anyhow.

Coincidentally, today I was discussing about how to document our standard properties and config options.

It really would be nice to be able to add some "official" way to add the configurable properties to the Sniffs docs (and, in parallel, maybe, be also able to have a "main" documentation file for the standard itself, allowing custom config options (and maybe also properties, if they are standard-wide).

The above is both a declaration and a question, heh, because I've not found any way to, officially, document them (other than manual docs). So I assume it doesn't exist (neither standard or sniff level).

Ciao :-)

**Comment by jrfnl** [comment]
@stronk7 You may be interested in having a read through #317...

**Comment by costdev** [comment]
Taking `Squiz.Commenting.BlockComment` as a first docs task for me.
